piqueserver/config.py

```diff
@@ -0,0 +1,86 @@
+import toml
+from pprint import pprint
+
+class ConfigException(Exception):
+    pass
+
+
+class ConfigEngine():
+    '''
+    configuration engine that handles loading and saving config,
+    and all the things
+    '''
+    def __init__(self):
+        self.raw_config = {}
+        self.style = 'TOML'
+
+    def load_config(self, config_file, style='TOML'):
+        # TODO: could also have this merge into current config if available
+        self.style = style
+
+        if style == 'TOML':
+            self.raw_config = toml.load(open(config_file))
+        else:
+            raise Exception()
+
+    def dump_config(self, out_file=None, style='TOML'):
+        if out_file is None:
+            pprint(self.raw_config)
+        # TODO
+        pass
+
+    def get(self, name, default):
+        # TODO: handle nested config values
+        # XXX: should requesting a config value that is a dictionary object
+        #       return the whole dictionary, or fail?
+        return self.raw_config.get(name, default)
+
+    def set(self, name, value):
+        self.raw_config[name] = value
+
+    def __call__(self, name, cast=None, default=None, validate=None):
+        return ConfigObject(self, name, cast, default, validate)
+
+
+
+class ConfigObject():
+    '''
+    configuration object stored in the engine
+    '''
+    def __init__(self, engine, name, cast, default, validate):
+        self.name = name
+        self.cast = cast
+        self.default = default
+        self.validate = validate
+        self.engine = engine
+
+        value = self.engine.get(name, default)
+        self.value = self._cast_and_validate(value)
+
+    def _cast_and_validate(self, value):
+        if self.cast is not None:
+            try:
+                value = self.cast(value)
+            except Exception as e:
+                print(e)
+                print('failed to cast config value')
+
+        if self.validate is not None:
+            ok = self.validate(value)
+            if not ok:
+                raise ConfigException('failed to validate')
+
+        return value
+
+
+    def get(self):
+        # TODO: use getattr magic to get and set this?
+        return self.value
+
+    def set(self, value):
```

Comment:
I'm not sure this  is a good idea at all... I'm worried that if we add this here, people might be tempted to use config values for storage. Besides, there is no really good reason to have this. The only part of the code that should be messing with config options directly is internal anyway.

Reply:
This was just for in case we wanted to make config files writeable, or to update runtime config. We need an api somewhere to update configuration.

Reply:
well, I'm saying there's no case in which you'd actually want to edit a setting through it's `ConfigObject`. In all places where you'd have a good reason to edit the global config, you have access to the global config already.

Reply:
What about in scripts or custom commands though?

Reply:
Script writers were exactly the people I feared would abuse this...